File: ZAJECIA_2/ZAD12/Dwukolorowanie.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

bool czy_dwukolorowy(vector<vector<int>> &graf, int liczba_wierzcholkow)
{
    vector<int> kolor(liczba_wierzcholkow, -1);
    for(int start = 0; start < liczba_wierzcholkow; start++)
    {
        if(kolor[start] == -1)
        {
            queue<int> kolejka;
            kolejka.push(start);
            //Kolorujemy pierwszym kolorem
            kolor[start] = 0;

            while(!kolejka.empty())
            {
                int wierzcholek = kolejka.front();
                kolejka.pop();
                for(int sasiad : graf[wierzcholek])
                {
                    //Jeszcze nie pokolorowany 
                    if(kolor[sasiad] == -1)
                    {
                        //Kolor przeciwny
                        kolor[sasiad] = 1 - kolor[wierzcholek];
                        kolejka.push(sasiad);
                    }
                    //Konflikt kolorow
                    else if(kolor[sasiad] == kolor[wierzcholek])
                    {
                        return false;
                    }
                }
            }
        }
    }
    return true;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int liczba_wierzcholkow, liczba_krawedzi;
    while(cin >> liczba_wierzcholkow) 
    {
        cin >> liczba_krawedzi;
        vector<vector<int>> graf(liczba_wierzcholkow);
        for(int i = 0; i < liczba_krawedzi; i++)
        {
            int u, v;
            cin >> u >> v;
            graf[u].push_back(v);
            graf[v].push_back(u);
        }
        if(czy_dwukolorowy(graf, liczba_wierzcholkow))
        {
            cout << "TAK";
        }
        else
        {
            cout << "NIE";
        }
        return 0;
    }
}

File: ZAJECIA_2/ZAD12/Dwukolorowanie.py
import sys
from collections import deque

def czy_dwukolorowy(graf, liczba_wierzcholkow):
    kolor = [-1] * liczba_wierzcholkow  
    
    for poczatek in range(liczba_wierzcholkow):
        #Jesli wierzcholek nie jest jeszcze pokolorowany
        if(kolor[poczatek] == -1):  
            kolejka = deque([poczatek])
            #Kolorujemy pierwszym kolorem
            kolor[poczatek] = 0  
            
            while kolejka:
                wierzcholek = kolejka.popleft()
                for sasiad in graf[wierzcholek]:
                    #Jeszcze nie pokolorowany
                    if(kolor[sasiad]) == -1:  
                        #Kolor przeciwny
                        kolor[sasiad] = 1 - kolor[wierzcholek]  
                        kolejka.append(sasiad)
                    #Konflikt kolorow
                    elif(kolor[sasiad] == kolor[wierzcholek]):  
                        return False
    return True

def main():
    liczba_wierzcholkow = int(sys.stdin.readline().strip())
    liczba_krawedzi = int(sys.stdin.readline().strip())

    graf = [[] for _ in range(liczba_wierzcholkow)]

    for _ in range(liczba_krawedzi):
        u, v = map(int, sys.stdin.readline().split())
        graf[u].append(v)
        graf[v].append(u) 

    if(czy_dwukolorowy(graf, liczba_wierzcholkow)):
        print("TAK")
    else:
        print("NIE")

if __name__ == "__main__":
    main()
